torch/csrc/jit/codegen/cuda/lower_validation.cpp

```diff
@@ -29,6 +32,180 @@ void validateIr(Fusion* fusion) {
+
+    TORCH_CHECK(
+        vector_size_optional.has_value(),
+        "Could not evalualte constant value bound to vectorized dim.");
+
+    auto vector_size =
+        dataTypeSize(tv->getDataType().value()) * vector_size_optional.value();
+
+    // Allow half2, float2, float4 and same sized vtypes.
+    std::array<int, 3> allowed_vector_sizes = {4, 8, 16};
+
+    TORCH_CHECK(
+        std::any_of(
+            allowed_vector_sizes.begin(),
+            allowed_vector_sizes.end(),
+            [&vector_size](int allowed_size) {
+              return vector_size == allowed_size;
+            }),
+        "Tried to vectorize a dim resulting in a word size of ",
+        vector_size,
+        " however, vector sizes only upto and including 16 bytes are supported.");
+
+    auto replay_exprs = ExprSort::getExprs(fusion, {v_id});
+
+    VectorizeValidator validator;
+
+    validator.vectorized_id = v_id;
+    for (auto expr_it = replay_exprs.rbegin(); expr_it != replay_exprs.rend();
+         ++expr_it) {
+      auto expr = *expr_it;
+      validator.handle(expr);
+    }
+
+    TORCH_CHECK(
+        validator.is_valid,
+        "Invalid vectorized pattern found, vectorization iter domains must be descendants of inner most dimension.",
+        "Issue found in, ",
+        tv);
+
+    TORCH_INTERNAL_ASSERT(validator.vectorized_id != nullptr);
+    auto root_dim = validator.vectorized_id;
+
+    // TODO: Contiguity is based on root domain not rfactor. Seems this
+    // generally doesn't cause problems, though contiguity should be on rfactor
+    // domain as that's the domain we index on.
+    IterDomain* last_root_dim = nullptr;
+    int last_root_dim_pos = -1;
+    for (size_t i = tv->getRootDomain().size(); i > 0; i--) {
+      auto r_id = tv->getRootDomain()[i - 1];
+      if (r_id->isReduction() || r_id->isBroadcast()) {
+        continue;
+      }
+      last_root_dim = r_id;
+      last_root_dim_pos = i - 1;
+      break;
+    }
+
+    if (last_root_dim == nullptr) {
+      // Should never get here, but that would mean there are no concrete dims,
+      // so we should be fine.
+      return;
+    }
+
+    TORCH_CHECK(
+        last_root_dim == validator.vectorized_id &&
+            tv->domain()->contiguity()[last_root_dim_pos],
+        "Vectorized dim has to be from a contiguous inner most position.");
+  }
+};
+
+} // namespace
+
+void validateVectorize(Fusion* fusion) {
+  FUSER_PERF_SCOPE("validateVectorize");
+  FusionGuard fg(fusion);
+
+  auto used_vals = DependencyCheck::getAllValsBetween(
+      {fusion->inputs().begin(), fusion->inputs().end()}, fusion->outputs());
+
+  std::unordered_set<TensorView*> used_tvs;
+
+  for (auto val : used_vals) {
+    if (ir_utils::isTV(val)) {
+      used_tvs.emplace(val->as<TensorView>());
+    }
+  }
+
+  for (auto tv : used_tvs) {
+    bool has_vectorize_dim = false;
+
+    for (size_t i = 0; i < tv->nDims(); i++) {
+      IterDomain* id = tv->axis(i);
+      IterDomain* concrete_id =
```

Comment:
nit: Is this necessary? Vectorization is only allowed to appear to the right of the computeAt position, so there should be no mapping in the parallel map.

Reply:
```
tv0->computeAt(tv1, 2);
tv1->axis(1)->parallelize(ParallelType::Vectorize);
```
Is illegal, but I think this is the only succinct way to catch it today.

Reply:
I got it. Thanks.